File: nfclib/src/main/java/com/ppy/nfclib/NfcStatusChangeBroadcastReceiver.java
package com.ppy.nfclib;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.nfc.NfcAdapter;

/**
 * Created by ZP on 2019/3/18.
 */
public class NfcStatusChangeBroadcastReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        String action = intent.getAction();
        if (NfcAdapter.ACTION_ADAPTER_STATE_CHANGED.equals(action)) {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN_MR2) {
                int status = intent.getIntExtra(NfcAdapter.EXTRA_ADAPTER_STATE, NfcAdapter.STATE_OFF);
                switch (status) {
                    case NfcAdapter.STATE_OFF:
                        onNfcOff();
                        break;
                    case NfcAdapter.STATE_ON:
                        onNfcOn();
                        break;
                    case NfcAdapter.STATE_TURNING_OFF:
                        onNfcTurningOff();
                        break;
                    case NfcAdapter.STATE_TURNING_ON:
                        onNfcTurningOn();
                        break;
                }
            }
        }
    }

    protected void onNfcTurningOn() {

    }

    protected void onNfcTurningOff() {

    }

    protected void onNfcOn() {

    }

    protected void onNfcOff() {

    }

    public static IntentFilter getNfcBroadcastReceiverIntentFilter() {
        return new IntentFilter("android.nfc.action.ADAPTER_STATE_CHANGED");
    }
}

File: nfclib/src/main/java/com/ppy/nfclib/NfcStatusChangeBroadcastReceiver.kt
package com.ppy.nfclib

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.nfc.NfcAdapter
import android.os.Build
import com.ppy.nfclib.util.Logger

/**
 * NFC开启、关闭系统广播.
 * Created by ZP on 2019/3/18.
 */
open class NfcStatusChangeBroadcastReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        val action = intent.action
        if (NfcAdapter.ACTION_ADAPTER_STATE_CHANGED == action) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                val state =
                    intent.getIntExtra(NfcAdapter.EXTRA_ADAPTER_STATE, NfcAdapter.STATE_OFF)

                Logger.get().println("nfc state broadcast receiver, state is $state")
                when (state) {
                    NfcAdapter.STATE_OFF -> onNfcOff()
                    NfcAdapter.STATE_ON -> onNfcOn()
                    NfcAdapter.STATE_TURNING_OFF -> onNfcTurningOff()
                    NfcAdapter.STATE_TURNING_ON -> onNfcTurningOn()
                    5 -> onCardPayMode()//samsumg return 5 that minds card pay mode
                    else -> onNfcOff()
                }
            }
        }
    }

    protected open fun onNfcTurningOn() {

    }

    protected open fun onNfcTurningOff() {

    }

    protected open fun onNfcOn() {

    }

    protected open fun onNfcOff() {

    }

    protected open fun onCardPayMode() {

    }

    companion object {

        val nfcBroadcastReceiverIntentFilter: IntentFilter
            get() = IntentFilter("android.nfc.action.ADAPTER_STATE_CHANGED")
    }
}
